addons/mail/controllers/discuss.py

```diff
@@ -420,10 +421,22 @@ def mail_thread_data(self, thread_model, thread_id, request_list, **kwargs):
 
     @http.route('/mail/thread/messages', methods=['POST'], type='json', auth='user')
     def mail_thread_messages(self, thread_model, thread_id, max_id=None, min_id=None, limit=30, **kwargs):
+        # Need a sudo because mail.mail has different ACL than mail.message
```

Comment:
I would not change that in stable ... Fetching messages is already a costly process, I fear that this kind of complex query will make it even slower. Anyway for 14 I would do the minimal patch to enable blacklist, and keep current behavior. As anyway people should already remove blacklisted emails from their marketing emails, I hope the extended usage of blacklist won't generate too much notifications. And if it does, well ... we will see :D .

What do you think @std-odoo ?

Reply:
I think the most expensive part of the domain is `('mail_ids', '=', False)`, but hard to say without a benchmark :/

Reply:
Following our discussion, for the message displayed on the client side, instead of the complex query in the controller that query the message and also remove undesired message (mail_bl, …), I have kept the simple query and added a simple one that search among the already filtered messages the one to exclude (mail_bl, …).
From the base version we have 3 queries instead of one but 2 are based on a very limited list of ids.
The added query generated by the framework (from the previous version) is:
```sql
SELECT "mail_message".id
  FROM "mail_message"
 WHERE (("mail_message"."id" in (119,118,117,116)) AND ("mail_message"."id" IN (SELECT "mail_mail"."mail_message_id" FROM "mail_mail" WHERE ("mail_mail"."failure_type" in ('mail_bl','mail_optout','mail_dup')))))
 ORDER BY "mail_message"."id" DESC
```
(where 119,118,117,116 is the id of the message related to a record)